drivers/can/can_tcan4x5x.c

```diff
@@ -0,0 +1,796 @@
+	__ASSERT_NO_MSG(len % 4 == 0);
+	__ASSERT_NO_MSG(len32 <= 256);
+
+	err = spi_transceive_dt(&tcan_config->spi, &tx, &rx);
+	if (err != 0) {
+		LOG_ERR("failed to read addr %u, len %d (err %d)", addr, len, err);
+		return err;
+	}
+
+	__ASSERT_NO_MSG((global_status & CAN_TCAN4X5X_IR_SPIERR) == 0U);
+
+	for (i = 0; i < len32; i++) {
+		dst32[i] = sys_be32_to_cpu(dst32[i]);
+	}
+
+	return 0;
+}
+
+static int tcan4x5x_write(const struct device *dev, uint16_t addr, const void *src, size_t len)
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct tcan4x5x_config *tcan_config = mcan_config->custom;
+	size_t len32 = len / sizeof(uint32_t);
+	uint32_t src32[len32];
+	uint8_t cmd[4] = {CAN_TCAN4X5X_WRITE_B_FL, addr >> 8U & 0xFF, addr & 0xFF,
+			  len32 == 256 ? 0U : len32};
+	uint8_t global_status;
+	const struct spi_buf tx_bufs[] = {
+		{.buf = &cmd, .len = sizeof(cmd)},
+		{.buf = &src32, .len = len},
+	};
+	const struct spi_buf rx_bufs[] = {
+		{.buf = &global_status, .len = sizeof(global_status)},
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_bufs,
+		.count = ARRAY_SIZE(tx_bufs),
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_bufs,
+		.count = ARRAY_SIZE(rx_bufs),
+	};
+	int err;
+	int i;
+
+	if (len == 0) {
+		return 0;
+	}
+
+	/* Maximum transfer size is 256 32-bit words */
+	__ASSERT_NO_MSG(len % 4 == 0);
+	__ASSERT_NO_MSG(len32 <= 256);
+
+	for (i = 0; i < len32; i++) {
+		src32[i] = sys_cpu_to_be32(((uint32_t *)src)[i]);
+	}
+
+	err = spi_transceive_dt(&tcan_config->spi, &tx, &rx);
+	if (err != 0) {
+		LOG_ERR("failed to write addr %u, len %d (err %d)", addr, len, err);
+		return err;
+	}
+
+	__ASSERT_NO_MSG((global_status & CAN_TCAN4X5X_IR_SPIERR) == 0U);
+
+	return 0;
+}
+
+static inline int tcan4x5x_read_tcan_reg(const struct device *dev, uint16_t reg, uint32_t *val)
+{
+	return tcan4x5x_read(dev, reg, val, sizeof(uint32_t));
+}
+
+static inline int tcan4x5x_write_tcan_reg(const struct device *dev, uint16_t reg, uint32_t val)
+{
+	return tcan4x5x_write(dev, reg, &val, sizeof(uint32_t));
+}
+
+static int tcan4x5x_read_mcan_reg(const struct device *dev, uint16_t reg, uint32_t *val)
+{
+	return tcan4x5x_read(dev, CAN_TCAN4X5X_MCAN_BASE + reg, val, sizeof(uint32_t));
+}
+
+static int tcan4x5x_write_mcan_reg(const struct device *dev, uint16_t reg, uint32_t val)
+{
+	return tcan4x5x_write(dev, CAN_TCAN4X5X_MCAN_BASE + reg, &val, sizeof(uint32_t));
+}
+
+static int tcan4x5x_read_mcan_mram(const struct device *dev, uint16_t offset, void *dst, size_t len)
+{
+	return tcan4x5x_read(dev, CAN_TCAN4X5X_MRAM_BASE + offset, dst, len);
+}
+
+static int tcan4x5x_write_mcan_mram(const struct device *dev, uint16_t offset, const void *src,
+				    size_t len)
+{
+	return tcan4x5x_write(dev, CAN_TCAN4X5X_MRAM_BASE + offset, src, len);
+}
+
+static int tcan4x5x_clear_mcan_mram(const struct device *dev, uint16_t offset, size_t len)
+{
+	static const uint8_t buf[256] = {0};
```

Comment:
This looks like it wastes a lot of flash memory (if the compiler isn't able to optimize it away).

This function is only called when a filter is removed and once during driver init. Maybe we could just use 4 zero-bytes on the stack instead of these 256? This would be sufficient for removal of the rx filter, but it would lead to longer initialization because of significantly more SPI transfers.

Just an idea, not blocking.

Reply:
The `tcan4x5x_clear_mcan_mram()` takes up 88 bytes of flash. The buffer is stack allocated. Clearing the MRAM in 4 bytes chunks would result in 512 individual SPI transactions. A buffer size of 256 bytes was chosen since this is the largest TCAN4x5x SPI transfer size.

```
$ make rom_report |grep tcan4x5x_clear_mcan_mram
    │   │       ├── tcan4x5x_clear_mcan_mram                                                       88   0.07%
```

Reply:
As the buffer is a static variable, it is not counted as part of the function, but appears at the top of the `rom_report`:

```
Root                                                                                            90116 100.00%
├── (hidden)                                                                                    20108  22.31%
├── (no paths)                                                                                   5696   6.32%
│   ...
│   ├── buf.0                                                                                     256   0.28%
```

Anyway, don't have any hard feelings about this. We can keep it as is.

Reply:
I would prefer a compromise with a buffer of 16 or 32 bytes.
The mram clear is only executed once.